### DISCUSSION TITLE
Gather ideas for CloudStack Extensions

### DISCUSSION THREAD

**Discussion by yadvr** [apache/cloudstack#9952] [Ideas]
Add here, your idea(s) which extend an existing CloudStack functionality, framework or feature that are simpler in nature and something an admin/sysadmin/sre/ops can write themselves in simpler shell/python scripts or a binary.

**Comment by taocnrut** [comment]
Let us run custom scripts when VM transitions from state to state: running, stopped, prepare for migration, snapshotting started/ended, etc. Also when specific alerts happen...

**Reply by DaanHoogland** [reply]
consider webhooks as implementation of this (4.20)

**Comment by taocnrut** [comment]
let us display aditional data inside cloudstack interface about VM, VPS, project... of a user by running executable that receives arguments and outputs ready html, some json, anything like that

**Reply by weizhouapache** [reply]
cloudstack has provided a command line which can support all cloudstack APIs and export the results in different formats
```
(localcloud) 🐱 > set output 
column  csv     json    table   text    default 
```
see https://github.com/apache/cloudstack-cloudmonkey

do you want the results to be displayed on UI ?

**Reply by taocnrut** [reply]
Well no, i was thinking more in direction of fetching some information about objects in cloudstack from data sources external to cloudstack and cloudstack would display it. So for example, cloudstack would pass VM name/uuid to a custom script and that script would return status of last external security scan of that VM.

**Reply by weizhouapache** [reply]
ok , understood @taocnrut 

actually we have a requirement to list external statistics, we ended up with a new UI plugin
since each user has different scripts, arguments and output format. would it make sense to write your own plugin (with API/UI/DB changes) ?

**Reply by taocnrut** [reply]
That is logical thing to do, wondering whether extensions could be used in a general way for something like that...

**Comment by phsm** [comment]
# Allow Cloudstack admins to customize libvirt domain XML.
Currently Cloudstack generates libvirt domain XML exclusively inside Java code.
Therefore, there little to no options to customize the KVM virtual machine settings. E.g. add some extra info into the metadata, or add some specific device etc.

It will make more sense to maintain the domain XML as a template somewhere in /usr/share/cloudstack/ that can be freely edited by the Cloudstack admins.
I imagine it being a Jinja (or something like [StringTemplate](https://www.stringtemplate.org/)) template that has a VM skeleton defined, and a bunch of template variables put into NICs, VM disk paths, disk drivers, VM name etc.

# Standardize and document the interface between Cloudstack and Virtual Routers.
This will allow Cloudstack admins to build their own custom VRs, e.g. a VR that uses something like [VPP](https://s3-docs.fd.io/vpp/25.02/) instead of iptables. 
I personally would like to experiment on making virtual routers more reliable and performant but this initiative shatters against having to sift through tens of thousands lines of Java code to figure out the interaction interface.

**Comment by weizhouapache** [comment]
> Standardize and document the interface between Cloudstack and Virtual Routers.
> This will allow Cloudstack admins to build their own custom VRs, e.g. a VR that uses something like [VPP](https://s3-docs.fd.io/vpp/25.02/) instead of iptables.
> I personally would like to experiment on making virtual routers more reliable and performant but this initiative shatters against having to sift through tens of thousands lines of Java code to figure out the interaction interface.

@phsm 
have you tried with VNF as a VR alternative ?  I know some users uses VNF instead of VR in their networks

We will maybe introduce a VNF framework in the next versions.
Users can implement for different router OSes (e.g pfsense, mikrotik, vyos)

**Reply by phsm** [reply]
Thanks Wei, I should definitely check VNF. 

But still, the interface between VR and CS isn't documented anywhere that I know of. If I knew how they interact with each other, I'd be able to contribute to the VR machine development.

**Comment by borisstoyanov** [comment]
introduce new type of systemVM called AIVM. This VM will analyze the cloudstack management logs, and answer possible resons for job failures, downtime and other troubleshooting.

**Comment by luganofer** [comment]
### 💻 Cloud Shell Service based on CloudMonkey
Cloud Shell Service based on CloudMonkey
Create a Cloud Shell service similar to the service provided by OCI (https://docs.oracle.com/es-ww/iaas/Content/API/Concepts/devcloudshellgettingstarted.htm) but based on the deployment of a new SystemVM with CloudMonkey pre-installed (cmk-vm 🐵 ) in which cloudmonkey is automatically configured with the user's apikey-secretkey and the CloudStack URL.  This way a user unfamiliar with cloudmonkey will have CLI access to the CloudStack APIs quickly and easily.  The VM with CloudMonkey can be a volatile VM such as VRs, internal LBs, etc.

**Reply by kiranchavala** [reply]
@luganofer Currently CMK  tool comes pre installed in the management server

**Reply by luganofer** [reply]
Hi @kiranchavala ...I mean using cmk as a service for users, not for CloudStack administrators (domain administrators and regular users do not have access to the managers). You can check the link on how to use the cloud shell service in OCI to get a better idea.

**Reply by PPisz** [reply]
It seems simple and unnecessary, but I consider it a very necessary functionality. I recently helped a Senior DevOps who knows cs kubernetes like a charm, but couldn't handle cloudmonkey ;-)

**Reply by PPisz** [reply]
I would go further, since cmk is part of mgmt, maybe we should add a command line to the gui? For admin level users, at the bottom of the screen. Like an extension of the "search" option at the top of the screen.

**Reply by luganofer** [reply]
> It seems simple and unnecessary, but I consider it a very necessary functionality. I recently helped a Senior DevOps who knows cs kubernetes like a charm, but couldn't handle cloudmonkey ;-)

Yes, I agree.  I have similar cases to the ones you are talking about.
If there was a pre-installed service and automated initial configuration of cloudmonkey, I think more users could take advantage of the use and benefits of CloudStack APIs via CLI access and directly start playing with cloudmonkey commands.

**Comment by kiranchavala** [comment]
Integrate Cloudstack with CRM tools (Jira, zendesk) for ticket creation on job failures 

Deploy vm via Slack

**Comment by PPisz** [comment]
VM (RBD) snapshot with RAM to RBD (RAM would be dumped to a separate volume). Vmware on vSAN has no problem with this...

**Reply by weizhouapache** [reply]
> VM (RBD) snapshot with RAM to RBD (RAM would be dumped to a separate volume). Vmware on vSAN has no problem with this...

good idea 
+1

**Comment by PPisz** [comment]
Another much-needed feature: https://github.com/apache/cloudstack/issues/9958

**Comment by luganofer** [comment]
📑 Adding CloudStack Events for internal audit purposes: [#9445](https://github.com/apache/cloudstack/issues/9445)

**Comment by pratikchandrakar** [comment]
> Add here, your idea(s) which extend an existing CloudStack functionality, framework or feature that are simpler in nature and something an admin/sysadmin/sre/ops can write themselves in simpler shell/python scripts or a binary.

Following features may be considered.
1. Searching VM with Nat IP. The default instances listing should show the Nat IP as well
2. Templates are hard linked with VMs, therefore old templates cannot be deleted. Storage wasting. Also, there is no option to update the existing template like VMware offers convert the image to VM and then to Image.
3. Offline deployment of Kubernetes cluster.
4. Alert improvement based on resource usage to both user and Admin.
